### PR TITLE
i#6685 core-shard: Add only_shards drmemtrace filter

### PR DESCRIPTION
**Pull request by derekbruening**
Adds a new filter by shard ordinal to the drmemtrace scheduler and analyzers, and adds multi-thread filtering on the command line.

Adds input_workload_t.only_shards to filter scheduler inputs by ordinal, which is mostly useful for core-sharded-on-disk traces.

Adds new CLI options -only_shards, which takes a comma-separated list of ordinals, and -only_threads, which takes a comma-separated list of tids.

Adds sorting of file names when reading inputs from a directory so that ordinals are reliable.  Adds leading 0's to record_filter's output name so this lexicographic sort keeps cores in order.

Adds error checking that any tid or ordinal filter is actually present in the inputs.

Adds an internal struct input_read_info_t to implement the new filtering inside the scheduler.

Adds scheduler unit tests of the new filters.

Adds an end-to-end test of -only_shards on a core-sharded-on-disk trace.

Also tested end-to-end manually with invalid tids, invalid ordinals, valid tids for core-sharded-on-disk, and valid ordinals for core-sharded-on-disk.

Issue: #6685

### CODE REVIEW

**Comment by derekbruening** [comment]
x32 failure is encoding size in opcode_mix #6303
a64-sve is sim.invariants #6733
a64 failure is startstop: searching for a prior issue; will file if new; re-running.

**Comment by derekbruening** [comment]
Filed #6928 for the api.startstop a64 failure.  This PR does not change core/ so it should have zero effect; should pass on retry.

File: clients/drcachesim/common/options.cpp [hunk 1]
```diff
@@ -545,7 +545,25 @@ droption_t<int>
...
+                   "thread with the given identifier.  0 enables all threads.  "
+                   "Applies only to -indir, not to -infile.  "
+                   "Cannot be combined with -only_threads or -only_shards.");
+
+droption_t<std::string>
+    op_only_threads(DROPTION_SCOPE_FRONTEND, "only_threads", "",
+                    "Only analyze these comma-separated threads",
+                    "Limits analyis to the list of comma-separated thread ids.  "
+                    "Applies only to -indir, not to -infile.  "
+                    "Cannot be combined with -only_shards.");
```
Comment (brettcoon) [inline clients/drcachesim/common/options.cpp:557]: Probably should say "Cannot be combined with -only_thread or -only_shards."

File: clients/drcachesim/common/options.cpp [hunk 2]
```diff
@@ -545,7 +545,25 @@ droption_t<int>
...
+droption_t<std::string>
+    op_only_shards(DROPTION_SCOPE_FRONTEND, "only_shards", "",
+                   "Only analyze these comma-separated shard ordinals",
+                   "Limits analyis to the list of comma-separated shard ordinals. "
+                   "A shard is typically an input thread but might be a core for "
+                   "core-sharded-on-disk traces.  The ordinal is 0-based and indexes "
+                   "into the sorted order of input filenames.  "
+                   "Applies only to -indir, not to -infile.  "
+                   "Cannot be combined with -only_threads.");
```
Comment (brettcoon) [inline clients/drcachesim/common/options.cpp:566]: Same comment about mentioning "-only_thread"

File: clients/drcachesim/scheduler/scheduler.h [hunk 3]
```diff
@@ -1490,14 +1514,13 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
     // Opens up all the readers for each file in 'path' which may be a directory.
```
Comment (brettcoon) [inline clients/drcachesim/scheduler/scheduler.h:1514]: I know it didn't change in this PR, but I think this comment could be clearer.
It doesn't mention that open_readers is selective about what it opens (subject to
only_threads and only_shards), and on my initial reading the first sentence seemed
to say it only opens readers for paths that might be directories.

How about something like "Opens readers for each file in 'path', subject to the
constraints in 'reader_info'.  'path' may be a directory."

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 4]
```diff
@@ -706,28 +706,136 @@ test_only_threads()
...
+                           scheduler_t::make_scheduler_serial_options(/*verbosity=*/4)) !=
+            scheduler_t::STATUS_SUCCESS)
+            assert(false);
+        auto *stream = scheduler.get_stream(0);
+        memref_t memref;
+        for (scheduler_t::stream_status_t status = stream->next_record(memref);
+             status != scheduler_t::STATUS_EOF; status = stream->next_record(memref)) {
+            assert(status == scheduler_t::STATUS_OK);
+            assert(memref.instr.tid == TID_B);
+        }
```
Comment (brettcoon) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:737]: Maybe add an assert after the loop that it actually read some non-STATUS_EOF records?

**Review by brettcoon — APPROVED** [review]
Couple suggestions, but looks good.

File: clients/drcachesim/analyzer.h [hunk 5]
```diff
@@ -216,9 +217,10 @@ template <typename RecordType, typename ReaderType> class analyzer_tmpl_t {
         operator=(const analyzer_worker_data_t &) = delete;
     };
 
-    // Pass INVALID_THREAD_ID for only_thread to include all threads.
     bool
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/analyzer.h:220]: Done.

File: clients/drcachesim/common/options.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/common/options.cpp:557]: Done.

File: clients/drcachesim/common/options.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/common/options.cpp:566]: Done.

File: clients/drcachesim/scheduler/scheduler.h (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:1514]: Done.  Similar for open_reader().  Also updated the Returns line to say Updates the ti2dinput, unfiltered_tids, and input_count fields of 'reader_info'..

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 4 above)
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:737]: Done.

**Pull request merged**